### ISSUE TITLE
nccl_net.h not found when installing via official Debian pkgs libnccl2 libnccl-dev

### ISSUE THREAD

**Issue by g-karthik** [NVIDIA/nccl#463]
Hello, I'm doing distributed training with AWS p3dn nodes (using Docker) and trying to set up Elastic Fabric Adapter (EFA) to increase node-node communication bandwidth.

In order to use EFA, I need to install the AWS OFI NCCL plugin available [here](https://github.com/aws/aws-ofi-nccl) into my Docker image. My image is based off an official Horovod base image located [here](https://hub.docker.com/layers/horovod/horovod/0.19.3-tf2.1.0-torch-mxnet1.6.0-py3.6-gpu/images/sha256-f675e5b16e90935526703acce06b55b78e43b2f53a0a69cafb1db1d064e13bef?context=explore). You can see in layers 19 and 21 that `libnccl2` (NCCL version 2.4.8) and `libnccl-dev` are properly pre-installed in this base image. I have been using this image successfully for my PyTorch distributed training jobs, so I do know that NCCL is set up properly.

However, when I try to install the AWS OFI NCCL plugin, it fails because it is unable to find `nccl_net.h`. In this GitHub repo containing the NCCL source, I do see that NCCL 2.4.8 has `nccl_net.h`. But why am I unable to find `nccl_net.h` in my Docker image? I only see `nccl.h`.

It appears that if an image installs NCCL from source (via this GitHub repo), it's better than installing NCCL via the Debian packages -- however, the latter is the officially recommended approach on the NVIDIA Developer docs.

Can you please advise on how I can unblock myself here with least damage to my Docker image and my downstream jobs?

Should I just install from source on top of these Debian packages? Will there be a conflict/errors in my downstream PyTorch jobs due to "multiple" NCCL installations, etc.?

@kwen2501 @sjeaugey

**Comment by g-karthik** [comment]
@sjeaugey any thoughts on the above?

**Comment by AddyLaddy** [comment]
@rashikakheria Can you give any advice to this user please?

**Comment by g-karthik** [comment]
@AddyLaddy I'm trying out the NCCL-from-source installation approach in parallel.

But this issue was more for me to understand why `nccl_net.h` is unavailable with the Debian packages, which is the officially recommended way to install NCCL on the NCCL installation [guide](https://docs.nvidia.com/deeplearning/nccl/install-guide/index.html). I don't think @rashikakheria would be able to help with answering that.

**Comment by AddyLaddy** [comment]
Our recommendation to plugin providers was to supply their own copy of the nccl_net.h header files as is done in the Mellanox plugin. I need to check on an AWS instance how they get around needing to include nccl_net.h apart from having access to a nccl source build tree.
We are planning to start releasing nccl_net.h in future packages but that won't happen until 2.9.x

**Comment by g-karthik** [comment]
@AddyLaddy does the Mellanox OFED and nv_peer_mem even work on AWS instances like p3dn and p4d? I thought EFA is the only workable solution on AWS for high-bandwidth.

**Comment by AddyLaddy** [comment]
No I don't believe it does. You need to use the EFA stack.
I was meantioning the MLNX plugin as it took a different approach to supplying the nccl_net.h header file - even before we bought the company.
On AWS I generally use one of the EFA enabled images that Amazon supply. In those images they have pre-built NCCL libraries and the corresponding headers e.g. 

>  find /usr/local/cuda-11.0/ -name '*nccl*'
/usr/local/cuda-11.0/targets/x86_64-linux/include/nccl.h
/usr/local/cuda-11.0/targets/x86_64-linux/include/nccl_net.h
/usr/local/cuda-11.0/lib/libnccl.so
/usr/local/cuda-11.0/lib/libnccl.so.2
/usr/local/cuda-11.0/lib/pkgconfig/nccl.pc
/usr/local/cuda-11.0/lib/libnccl.so.2.7.8
/usr/local/cuda-11.0/lib/libnccl_static.a
/usr/local/cuda-11.0/efa/lib/libnccl-net.so.0.0.0
/usr/local/cuda-11.0/efa/lib/libnccl-net.so.0
/usr/local/cuda-11.0/efa/lib/libnccl-net.so
/usr/local/cuda-11.0/efa/lib/libnccl-net.la
/usr/local/cuda-11.0/efa/share/aws-ofi-nccl

**Comment by rashikakheria** [comment]
Currently, Deep learning AMIs build from source and put the `nccl_net.h` for users in a specified directory. For any user who would like to have a custom build of nccl is advised to build from source.

I think here the use-case of @g-karthik is to install nccl in a docker container without building from source (so they are not able to leverage the pre-packaged NCCL that we have inside our DL AMIs)

**Comment by g-karthik** [comment]
> Our recommendation to plugin providers was to supply their own copy of the nccl_net.h header files as is done in the Mellanox plugin. I need to check on an AWS instance how they get around needing to include nccl_net.h apart from having access to a nccl source build tree.
> We are planning to start releasing nccl_net.h in future packages but that won't happen until 2.9.x

@AddyLaddy Is `nccl_net.h` in now that 2.9.x has been released?

**Comment by AddyLaddy** [comment]
Yes `nccl_net.h` is now included in the NCCL 2.9.6 and future `libnccl-dev` packages.

**Comment by g-karthik** [comment]
@AddyLaddy great! so I presume folks wanting to use EFA with NCCL 2.9.6+ wouldn't need to install NCCL from source again on top of a base image that already has NCCL 2.9.6 installed, such as this one? https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_21-04.html#rel_21-04

@rashikakheria could you also confirm?

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
